File: day12/solution.go
// Advent of Code 2025 - Day 12: Christmas Tree Farm (polyomino packing)
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// Cell represents a position in a shape or grid.
// Go doesn't have tuples, so we use a struct instead.
type Cell struct {
	Row, Col int
}

// Shape is a set of cells. We use map[Cell]struct{} as Go's idiomatic "set" -
// struct{} takes zero bytes, so it's more memory efficient than map[Cell]bool.
type Shape map[Cell]struct{}

// Region holds the dimensions and required shape counts for a tree region
type Region struct {
	Width, Height int
	Counts        []int
}

// normalize shifts a shape so its top-left is at (0,0)
func normalize(shape Shape) Shape {
	if len(shape) == 0 {
		return shape
	}

	// find min row and col
	minRow, minCol := 1000, 1000
	for cell := range shape {
		if cell.Row < minRow {
			minRow = cell.Row
		}
		if cell.Col < minCol {
			minCol = cell.Col
		}
	}

	// shift everything
	result := make(Shape)
	for cell := range shape {
		result[Cell{cell.Row - minRow, cell.Col - minCol}] = struct{}{}
	}
	return result
}

// rotate90 rotates a shape 90 degrees clockwise: (r, c) -> (c, -r)
func rotate90(shape Shape) Shape {
	result := make(Shape)
	for cell := range shape {
		result[Cell{cell.Col, -cell.Row}] = struct{}{}
	}
	return normalize(result)
}

// flipHorizontal mirrors a shape: (r, c) -> (r, -c)
func flipHorizontal(shape Shape) Shape {
	result := make(Shape)
	for cell := range shape {
		result[Cell{cell.Row, -cell.Col}] = struct{}{}
	}
	return normalize(result)
}

// shapeToString converts a shape to a string for use as a map key.
// Go maps can't use slices or maps as keys, so we serialize to compare shapes.
func shapeToString(shape Shape) string {
	// collect cells and sort them for consistent ordering
	cells := make([]Cell, 0, len(shape))
	for cell := range shape {
		cells = append(cells, cell)
	}

	// simple bubble sort - shapes are tiny (7 cells max)
	for i := 0; i < len(cells); i++ {
		for j := i + 1; j < len(cells); j++ {
			if cells[i].Row > cells[j].Row ||
				(cells[i].Row == cells[j].Row && cells[i].Col > cells[j].Col) {
				cells[i], cells[j] = cells[j], cells[i]
			}
		}
	}

	var sb strings.Builder
	for _, c := range cells {
		fmt.Fprintf(&sb, "%d,%d;", c.Row, c.Col)
	}
	return sb.String()
}

// getAllOrientations returns all unique orientations (up to 8) of a shape
func getAllOrientations(shape Shape) []Shape {
	seen := make(map[string]struct{})
	var orientations []Shape

	current := shape
	for i := 0; i < 4; i++ {
		// try normal and flipped
		for _, s := range []Shape{current, flipHorizontal(current)} {
			normalized := normalize(s)
			key := shapeToString(normalized)
			if _, exists := seen[key]; !exists {
				seen[key] = struct{}{}
				orientations = append(orientations, normalized)
			}
		}
		current = rotate90(current)
	}

	return orientations
}

// parseShape converts lines like "###", "##.", ".##" into a Shape
func parseShape(lines []string) Shape {
	shape := make(Shape)
	for row, line := range lines {
		for col, ch := range line {
			if ch == '#' {
				shape[Cell{row, col}] = struct{}{}
			}
		}
	}
	return shape
}

// parseInput reads the puzzle input and returns shapes (with all orientations) and regions
func parseInput(filename string) (map[int][]Shape, []Region, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, nil, err
	}
	defer file.Close()

	shapes := make(map[int][]Shape)
	var regions []Region

	scanner := bufio.NewScanner(file)
	var currentLines []string
	var currentShapeIdx int = -1

	for scanner.Scan() {
		line := scanner.Text()

		// blank line ends current shape definition
		if line == "" {
			if currentShapeIdx >= 0 && len(currentLines) > 0 {
				shape := parseShape(currentLines)
				shapes[currentShapeIdx] = getAllOrientations(shape)
				currentLines = nil
				currentShapeIdx = -1
			}
			continue
		}

		// shape definition starts with "N:"
		if len(line) >= 2 && line[len(line)-1] == ':' {
			idx, err := strconv.Atoi(line[:len(line)-1])
			if err == nil {
				currentShapeIdx = idx
				currentLines = nil
				continue
			}
		}

		// region definition has "x" in dimensions
		if strings.Contains(line, "x") && strings.Contains(line, ": ") {
			parts := strings.SplitN(line, ": ", 2)
			dims := strings.Split(parts[0], "x")
			width, _ := strconv.Atoi(dims[0])
			height, _ := strconv.Atoi(dims[1])

			countStrs := strings.Fields(parts[1])
			counts := make([]int, len(countStrs))
			for i, s := range countStrs {
				counts[i], _ = strconv.Atoi(s)
			}

			regions = append(regions, Region{width, height, counts})
			continue
		}

		// otherwise it's part of a shape definition
		if currentShapeIdx >= 0 {
			currentLines = append(currentLines, line)
		}
	}

	// handle last shape if file doesn't end with blank line
	if currentShapeIdx >= 0 && len(currentLines) > 0 {
		shape := parseShape(currentLines)
		shapes[currentShapeIdx] = getAllOrientations(shape)
	}

	return shapes, regions, scanner.Err()
}

// Grid tracks which cells are occupied
type Grid struct {
	cells  [][]bool
	Height int
	Width  int
}

func newGrid(height, width int) *Grid {
	cells := make([][]bool, height)
	for i := range cells {
		cells[i] = make([]bool, width)
	}
	return &Grid{cells, height, width}
}

// canPlace checks if shape fits at (startRow, startCol) without overlapping
func (g *Grid) canPlace(shape Shape, startRow, startCol int) bool {
	for cell := range shape {
		r, c := startRow+cell.Row, startCol+cell.Col
		if r < 0 || r >= g.Height || c < 0 || c >= g.Width {
			return false
		}
		if g.cells[r][c] {
			return false
		}
	}
	return true
}

// place marks cells as occupied (fill=true) or empty (fill=false)
func (g *Grid) place(shape Shape, startRow, startCol int, fill bool) {
	for cell := range shape {
		g.cells[startRow+cell.Row][startCol+cell.Col] = fill
	}
}

// shapeInstance pairs a shape index with its possible orientations
type shapeInstance struct {
	idx          int
	orientations []Shape
}

// canPlaceAll uses backtracking to try placing all shapes on the grid
func canPlaceAll(toPlace []shapeInstance, grid *Grid) bool {
	// base case - everything placed!
	if len(toPlace) == 0 {
		return true
	}

	// grab next shape to place
	current := toPlace[0]
	remaining := toPlace[1:]

	// try each orientation at each position
	for _, orientation := range current.orientations {
		for row := 0; row < grid.Height; row++ {
			for col := 0; col < grid.Width; col++ {
				if grid.canPlace(orientation, row, col) {
					grid.place(orientation, row, col, true)

					if canPlaceAll(remaining, grid) {
						return true
					}

					// backtrack
					grid.place(orientation, row, col, false)
				}
			}
		}
	}

	return false
}

// canFitRegion checks if all required shapes fit in the region
func canFitRegion(shapes map[int][]Shape, region Region) bool {
	// quick check: enough cells?
	totalCells := 0
	for idx, count := range region.Counts {
		if count > 0 {
			if orientations, ok := shapes[idx]; ok && len(orientations) > 0 {
				totalCells += count * len(orientations[0])
			}
		}
	}

	if totalCells > region.Width*region.Height {
		return false
	}

	// build list of shapes to place
	var toPlace []shapeInstance
	for idx, count := range region.Counts {
		for i := 0; i < count; i++ {
			toPlace = append(toPlace, shapeInstance{idx, shapes[idx]})
		}
	}

	grid := newGrid(region.Height, region.Width)
	return canPlaceAll(toPlace, grid)
}

func main() {
	shapes, regions, err := parseInput("day12/input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	// Part 1: count regions that can fit all presents
	count := 0
	for _, region := range regions {
		if canFitRegion(shapes, region) {
			count++
		}
	}

	fmt.Printf("Part 1: %d\n", count)
}

File: day12/solution.py
"""Day 12: Christmas Tree Farm - polyomino packing puzzle."""

from pathlib import Path


def parse_shape(lines: list[str]) -> set[tuple[int, int]]:
    """Convert a shape's visual representation to a set of (row, col) coordinates."""
    cells = set()
    for row, line in enumerate(lines):
        for col, char in enumerate(line):
            if char == "#":
                cells.add((row, col))
    return cells


def normalize(shape: set[tuple[int, int]]) -> frozenset[tuple[int, int]]:
    """Shift shape so its top-left corner is at (0, 0).

    This makes shapes comparable regardless of where they started.
    """
    if not shape:
        return frozenset()
    min_row = min(r for r, c in shape)
    min_col = min(c for r, c in shape)
    return frozenset((r - min_row, c - min_col) for r, c in shape)


def rotate_90(shape: set[tuple[int, int]]) -> set[tuple[int, int]]:
    """Rotate shape 90 degrees clockwise.

    The trick: (row, col) -> (col, -row)
    Then we normalize to bring it back to origin.
    """
    return {(c, -r) for r, c in shape}


def flip_horizontal(shape: set[tuple[int, int]]) -> set[tuple[int, int]]:
    """Flip shape horizontally (mirror across vertical axis)."""
    return {(r, -c) for r, c in shape}


def get_all_orientations(shape: set[tuple[int, int]]) -> list[frozenset[tuple[int, int]]]:
    """Generate all unique orientations of a shape (up to 8).

    4 rotations × 2 (original + flipped) = 8 max, but some shapes
    have symmetry so we use a set to dedupe.
    """
    orientations = set()
    current = shape

    # try all 4 rotations
    for _ in range(4):
        orientations.add(normalize(current))
        orientations.add(normalize(flip_horizontal(current)))
        current = rotate_90(current)

    return list(orientations)


def parse_input(text: str) -> tuple[dict[int, list[frozenset]], list[tuple[int, int, list[int]]]]:
    """Parse the full input into shapes and regions.

    Returns:
        shapes: dict mapping shape index -> list of all orientations
        regions: list of (width, height, [counts for each shape])
    """
    sections = text.strip().split("\n\n")

    shapes = {}
    regions = []

    for section in sections:
        lines = section.strip().split("\n")

        # check if this is a shape definition (starts with "N:")
        if lines[0].rstrip(":").isdigit() or (lines[0].endswith(":") and lines[0][:-1].isdigit()):
            # shape definition
            idx = int(lines[0].rstrip(":"))
            shape_cells = parse_shape(lines[1:])
            shapes[idx] = get_all_orientations(shape_cells)

        elif "x" in lines[0]:
            # region definitions - could be multiple lines
            for line in lines:
                parts = line.split(": ")
                dims = parts[0].split("x")
                width, height = int(dims[0]), int(dims[1])
                counts = list(map(int, parts[1].split()))
                regions.append((width, height, counts))

    return shapes, regions


def can_place_at(grid: list[list[bool]], shape: frozenset[tuple[int, int]],
                 start_row: int, start_col: int, height: int, width: int) -> bool:
    """Check if we can place a shape at the given position."""
    for dr, dc in shape:
        r, c = start_row + dr, start_col + dc
        # out of bounds?
        if r < 0 or r >= height or c < 0 or c >= width:
            return False
        # already occupied?
        if grid[r][c]:
            return False
    return True


def place_shape(grid: list[list[bool]], shape: frozenset[tuple[int, int]],
                start_row: int, start_col: int, fill: bool):
    """Place or remove a shape from the grid."""
    for dr, dc in shape:
        grid[start_row + dr][start_col + dc] = fill


def can_place_all_shapes(shapes_to_place: list[tuple[int, list[frozenset]]],
                         grid: list[list[bool]], height: int, width: int) -> bool:
    """Backtracking search to place all shapes on the grid."""
    # base case - all shapes placed successfully!
    if not shapes_to_place:
        return True

    # grab the next shape to place
    shape_idx, orientations = shapes_to_place[0]
    remaining = shapes_to_place[1:]

    # try each orientation of this shape
    for orientation in orientations:
        # try placing at each grid position
        for row in range(height):
            for col in range(width):
                if can_place_at(grid, orientation, row, col, height, width):
                    # place it and recurse
                    place_shape(grid, orientation, row, col, True)

                    if can_place_all_shapes(remaining, grid, height, width):
                        return True

                    # didn't work - backtrack by removing the shape
                    place_shape(grid, orientation, row, col, False)

    # no valid placement found for this shape
    return False


def can_fit_region(shapes: dict[int, list[frozenset]],
                   width: int, height: int, counts: list[int]) -> bool:
    """Check if all required shapes can fit in the region."""
    # quick check: do we even have enough cells?
    total_cells_needed = 0
    for idx, count in enumerate(counts):
        if count > 0 and idx in shapes:
            # all orientations have same cell count, just grab first
            total_cells_needed += count * len(shapes[idx][0])

    if total_cells_needed > width * height:
        return False

    # build list of shapes we need to place (one entry per shape instance)
    shapes_to_place = []
    for idx, count in enumerate(counts):
        for _ in range(count):
            shapes_to_place.append((idx, shapes[idx]))

    # start with empty grid
    grid = [[False] * width for _ in range(height)]

    return can_place_all_shapes(shapes_to_place, grid, height, width)


def main():
    text = (Path(__file__).parent / "input.txt").read_text()
    shapes, regions = parse_input(text)

    # Part 1: count regions that can fit all their presents
    count = 0
    for width, height, shape_counts in regions:
        if can_fit_region(shapes, width, height, shape_counts):
            count += 1

    print(f"Part 1: {count}")


if __name__ == "__main__":
    main()
